### PR TITLE
Add user function definition support

### PR DESCRIPTION
**Pull request by lenticularis39**
A new structure for defining subprograms is added on the same level as probes:
```
fn <function_name>(<type0> <arg0>, ...): <return_type> {
  <body>
}
```
The existing return statement is extended to also allow returning a specified value. SemanticAnalyser checks if the type is correct, while CodegenLLVM goes through the generated code to ensure all code flow paths return a value in a non-void function.

Only a limited number of builtins are allowed to be used inside functions due to current implementation restrictions; these can be addressed in later development.

----

This is the first step in implementing #2516 that can be either merged separately or together with the rest of the implementation. Note that this does not implement a way to call the functions nor a way to generate the correct BPF code for that.

(I'm opening this draft PR, since I believe there is quite a few non-trivial changes that it makes sense to review while I work on the rest.)

##### Checklist

Note: this checklist is only for the function definition part, excluding the BPF code generator and function calls.

- [x] Tests for parser
- [x] Tests for semantic analyser
- [x] Tests for codegen
- [x] Function argument support

### CODE REVIEW

File: tests/parser.cpp
Comment (viktormalik) [inline tests/parser.cpp:1978]: Nit: would it be possible to add `()` here so that it is visible that it's a function?
```suggestion
       " f1: void()\n");
```

File: src/ast/passes/codegen_llvm.cpp
Comment (viktormalik) [inline src/ast/passes/codegen_llvm.cpp:2512]: This piece of code would deserve to be enclosed in a separate function, e.g. `checkMissingReturns` or so.

**Review by viktormalik — COMMENTED** [review]
Looks great, the implementation is pretty straightforward. Just a couple of comments.

I'd advocate for merging this before implementing support for function calls.

**Comment by lenticularis39** [comment]
Resolved comments and rebased.

File: src/ast/passes/codegen_llvm.cpp
Comment (viktormalik) [reply src/ast/passes/codegen_llvm.cpp]: This still applies, right?

Comment (lenticularis39) [reply src/ast/passes/codegen_llvm.cpp]: Ah yes, I somehow missed that you are suggesting to run unreachable code elimination. The codegen already generates dummy returns for probes, they are just simpler, since the return value of probes is either void or some kind of integer, so there is no need for an empty alloca.

On the other hand, running dead code elimination here might slightly impact the tests (e.g. the pass might fail on invalid LLVM IR or errors in unreachable code won't be caught in verification). I will try to put the pass there and see what it does.

Comment (lenticularis39) [reply src/ast/passes/codegen_llvm.cpp]: I tried running `UnreachableBlockElimPass` at the end of `void CodegenLLVM::visit(Subprog &subprog)` and it looks good. I think it might make sense to do that in a separate PR together with the same pass for probes rather than here, as I wrote above, this is not an issue specific to user-defined functions, it is just the most visible there.

File: src/ast/visitors.cpp
Comment (ajor) [inline src/ast/visitors.cpp:1]: I think Mutator would need to visit jump.return_value to be fully correct.
Doesn't it also need to visit subprog.args?

However, it seems like Mutator & Dispatcher aren't used anywhere, so this is all dead code anyway. This also means we can't test it.

I don't want to merge in potentially incorrect code, so I think we'd be better off deleting these classes entirely.

cc @fbs

File: src/lexer.l
Comment (ajor) [inline src/lexer.l:216]: Why are the space changes needed in lexer.l (this line and also in the `AFTER_COLON` section)? Why is space handled differently in this section?

**Review by ajor — CHANGES_REQUESTED** [review]
Love the tests!

The one main structural comment I've got is that something like checking all paths return a value should come under semantic analysis rather than codegen. Codegen should be able to assume that what it's been given is valid, and not worry about checking things itself.

Slightly off topic: We need to start caring more about the performance of the parser. Our approach previously was to not worry because it's not on the tracing hot-path, but the unit tests taking multiple seconds to run indicates that we've taken this too far. I did a quick profile and we're spending a lot of our time just copying strings around. I left a few comments on the new changes here, but we'll need to revisit the rest of ast.h at some point.

**Comment by danobi** [comment]
Going to set aside some time this weekend to take a look

File: tests/parser.cpp
Comment (lenticularis39) [reply tests/parser.cpp:1974]: Oops, I missed that.

File: src/ast/ast.h
Comment (lenticularis39) [reply src/ast/ast.h:554]: Ah, that is true, and I know that - no idea why I put the line there, might have been copied from somewhere and changed because of compiler warnings. Thanks!

File: src/ast/ast.h
Comment (lenticularis39) [reply src/ast/ast.h:555]: I agree, that was the intention - I was just focusing more on the functional side of things and forgot to change it.

File: src/lexer.l
Comment (lenticularis39) [reply src/lexer.l:216]: `AFTER_COLON` is an `%x` section, IIUC the usual rules do not apply there. Here the space handling should be a part of the logic ensuring `struct test $x` (for example) is parsed correctly.

I can take a closer look at it if that did not make much sense, that's what I remember from writing this code.

File: src/ast/visitors.cpp
Comment (lenticularis39) [reply src/ast/visitors.cpp:1]: I'm not sure myself what the Mutator class is supposed to do exactly, you are probably right.

**Comment by lenticularis39** [comment]
> Love the tests!

Thank you! I had to write thorough tests, the area of possibility of bugs here is almost unlimited :D

> 
> The one main structural comment I've got is that something like checking all paths return a value should come under semantic analysis rather than codegen. Codegen should be able to assume that what it's been given is valid, and not worry about checking things itself.
> 

I also thought that initially, however, checking if all paths return a value is actually not semantic analysis, but control flow analysis, since semantic analysis operates on the *AST*, while this checks for a property of the *CFG*. Implementing it in semantic analyser would require to construct the control flow inside the AST, which would duplicate work that is already done in codegen.

> Slightly off topic: We need to start caring more about the performance of the parser. Our approach previously was to not worry because it's not on the tracing hot-path, but the unit tests taking multiple seconds to run indicates that we've taken this too far. I did a quick profile and we're spending a lot of our time just copying strings around. I left a few comments on the new changes here, but we'll need to revisit the rest of ast.h at some point.

Good point.

File: src/ast/visitors.cpp
Comment (ajor) [reply src/ast/visitors.cpp:1]: After thinking about this some more and discussing the Mutator with @fbs, I think we're going to remove it entirely for now as it's not used. (it can be brought back when needed)

Basically: don't worry about the mutator in this PR

Comment (lenticularis39) [reply src/ast/visitors.cpp:1]: OK, thanks.

**Comment by ajor** [comment]
> I also thought that initially, however, checking if all paths return a value is actually not semantic analysis, but control flow analysis, since semantic analysis operates on the _AST_, while this checks for a property of the _CFG_. Implementing it in semantic analyser would require to construct the control flow inside the AST, which would duplicate work that is already done in codegen.

In languages with more complex control flow (e.g. `goto` statements) I think this makes sense, but is it necessary for bpftrace, which only has structured control flow?

I feel like something similar to this pseudo-code should be able to detect missing return statements:
```
fn all_paths_return(block):
  for stmt in block:
    if stmt == "return":
      return true
    if stmt == "break" or stmt == "continue":
      return false
    if stmt == "if":
      if all_paths_return(stmt.true_branch) && all_paths_return(stmt.false_branch):
        return true
    if stmt has children: (e.g. a loop)
      if all_paths_return(stmt.children):
        return true
  return false
```

File: src/ast/visitors.cpp
Comment (fbs) [reply src/ast/visitors.cpp:1]: goal of the mutator is to write ast modifying passes, it mutates one ast into another. But never got around to writing mutating passes

File: src/ast/ast.h
Comment (danobi) [inline src/ast/ast.h:665]: nit: maybe `return_type` instead of `type`? To differentiate between type of function and return type

File: src/ast/passes/codegen_llvm.h
Comment (danobi) [inline src/ast/passes/codegen_llvm.h:244]: nit: just to make it explicit
```suggestion
  bool inside_subprog_ = false;
```

File: src/ast/passes/codegen_llvm.cpp
Comment (danobi) [inline src/ast/passes/codegen_llvm.cpp:2631]: Same comment as @ajor -- doesn't feel right to do this in codegen

File: src/ast/passes/codegen_llvm.cpp
Comment (danobi) [inline src/ast/passes/codegen_llvm.cpp:2567]: Does moving return checks into semantic analyser get rid of the ret_dummy stuff? Not immediately clear to me what this does

File: tests/semantic_analyser.cpp
Comment (danobi) [inline tests/semantic_analyser.cpp:2715]: Sorry if this was discussed elsewhere already, but the argument syntax feels a bit odd to me.

The function syntax is `fn <idents and stuff>: <ret type>`. It seems to me it would follow that argument syntax should also be `<ident>: <type>`, but here it's reversed.

IIRC we also added type annotations for local variables as `$var: type = <blah>` right?

Comment (danobi) [reply tests/semantic_analyser.cpp:2715]: In that case I am in favor of `$arg: type` syntax

**Comment by fbs** [comment]
> > I also thought that initially, however, checking if all paths return a value is actually not semantic analysis, but control flow analysis, since semantic analysis operates on the _AST_, while this checks for a property of the _CFG_. Implementing it in semantic analyser would require to construct the control flow inside the AST, which would duplicate work that is already done in codegen.
> 
> In languages with more complex control flow (e.g. `goto` statements) I think this makes sense, but is it necessary for bpftrace, which only has structured control flow?
> 
> I feel like something similar to this pseudo-code should be able to detect missing return statements:
> 
> ```
> fn all_paths_return(block):
>   for stmt in block:
>     if stmt == "return":
>       return true
>     if stmt == "break" or stmt == "continue":
>       return false
>     if stmt == "if":
>       if all_paths_return(stmt.true_branch) && all_paths_return(stmt.false_branch):
>         return true
>     if stmt has children: (e.g. a loop)
>       if all_paths_return(stmt.children):
>         return true
>   return false
> ```

I'd put something like this in its own pass. Is not semantic but also shouldn't go in the codegen. Passes should be fairly easy to add as all the boilerplate is handled (see e.g. nodecount.h).

Might be a good usecase for the `Dispatcher`  visitor as it allows for a return value which makes it a bit easier to write. 

Might be good to do it in a separte PR to avoid some noise in this one

**Comment by lenticularis39** [comment]
> I'd put something like this in its own pass. Is not semantic but also shouldn't go in the codegen. Passes should be fairly easy to add as all the boilerplate is handled (see e.g. nodecount.h).
> 
> Might be a good usecase for the `Dispatcher` visitor as it allows for a return value which makes it a bit easier to write.
> 
That sounds like the best idea to me, that will avoid the check being in codegen and having it returning bool, and also complicating the code of semantic analyzer.

Also, there is a direct benefit of doing it on the AST as  @ajor and @danobi suggested: we can print the exact place where a return instruction should be and isn't, which will make it easier for the programmer.

> Might be good to do it in a separte PR to avoid some noise in this one
There might be just an issue with the fact that without the check, bpftrace will generate invalid LLVM IR; however, I don't think that is that big of an issue, since the function feature is not officially supported yet. I can also make it a separate commit, that might be better.

**Comment by fbs** [comment]
The idea behind the passmanager and all the visitor stuff is that it should make new passes easy to add, to avoid increasing the complexity of existing passes, the semantic analyser is quite tricky in places.

If you do want to try the dispather visitor, it will be the first usecase and its a bit hacky so it might not work that  well :(


Also, really nice work here :)

**Comment by lenticularis39** [comment]
Thank you all for the helpful comments, I feel like it cleared a lot of things up and will increase the quality of the PR a lot.

File: tests/semantic_analyser.cpp
Comment (fbs) [reply tests/semantic_analyser.cpp:2715]: I wonder if this type syntax can be expanded to other cases, like:

```
kprobe:some_kernel_fn(a: int, b: int): int
```

Basically adding the option to specify function signatures in places (althoug with btf and dwarf that might be a bit late). In that case, i wonder if having `->` like rust,python will make parsing easier

**Comment by lenticularis39** [comment]
Resolved (hopefully) all of the comments. There are some minor issues with the code like leftover declarartions, I will clean it after rebasing to current master.

**Comment by danobi** [comment]
Top of my todo list to review

File: tests/return_path_analyser.cpp
Comment (viktormalik) [inline tests/return_path_analyser.cpp:61]: Can we do something with the ugly linebreaks or is it clang-format enforcing this?
```suggestion
  test("fn test($x: int64): int64 {"
       "  if ($x > 0) { return 0; } else { return 1; }"
       "}",
```
Same below.

**Review by viktormalik — COMMENTED** [review]
Looks good! I mostly have some minor suggestions.

File: src/ast/passes/return_path_analyser.cpp
Comment (lenticularis39) [reply src/ast/passes/return_path_analyser.cpp:14]: Not only this file, it's a common ugliness of my code :D

File: tests/return_path_analyser.cpp
Comment (lenticularis39) [reply tests/return_path_analyser.cpp:61]: I don't think clang-format is enforcing this, I just let it auto-format the file and it did it like that.

File: src/ast/ast.cpp
Comment (lenticularis39) [reply src/ast/ast.cpp:234]: Ah I see, there is no double free, since I set `stmts` to `nullptr` so I didn't notice. Nice catch!

**Comment by danobi** [comment]
Took a close look at ast, parser, lexer, return_path_analyser. They look really nice. Will try to make another pass tmr morning over semantic analyser and codegen details.

File: src/ast/passes/semantic_analyser.cpp
Comment (danobi) [inline src/ast/passes/semantic_analyser.cpp:2205]: Nit: easier to read if `is_void && (return_value != nullptr)` IMO

**Review by danobi — COMMENTED** [review]
semantic analysis lgtm; one small nit

File: src/ast/passes/codegen_llvm.cpp
Comment (danobi) [inline src/ast/passes/codegen_llvm.cpp:2607]: Would it make sense to hoist this out to a helper function? I think `generateProbe()` shares the `variables_.clear()` bit. It would help self-document the intent and make future extensions less error prone

File: src/ast/passes/codegen_llvm.cpp
Comment (danobi) [inline src/ast/passes/codegen_llvm.cpp:2617]: Is it necessary to put arguments on stack? I know `variables_` value is an allocainst. But I'm not sure it's strictly necessary (unless I'm mistaken which I could easily be). AFAIU should be OK to keep variables in register.

File: src/ast/passes/codegen_llvm.h
Comment (lenticularis39) [reply src/ast/passes/codegen_llvm.h:171]: Subprogs are allowed to return a value, the comment only refers to `createRet()` being called without an argument/with null argument.

File: src/ast/passes/codegen_llvm.cpp
Comment (lenticularis39) [reply src/ast/passes/codegen_llvm.cpp:2607]: Parts of it are shared, but parts of it are also different, subprogs process arguments, probes do not, probe do additional setup subprogs don't.

**Review by danobi — APPROVED** [review]

File: src/ast/passes/codegen_llvm.h
Comment (danobi) [reply src/ast/passes/codegen_llvm.h:171]: Ah yeah. Misread the comment

**Comment by lenticularis39** [comment]
Just fixed a small readability issue (moving check for void return value to `ReturnPathAnalyser::visit(Subprog &subprog)`).

File: src/ast/passes/return_path_analyser.cpp
Comment (lenticularis39) [reply src/ast/passes/return_path_analyser.cpp]: Fixed.

**Review by viktormalik — APPROVED** [review]
Thanks!

@ajor wants to review this, too, so let's wait for that before merging.

File: src/ast/passes/semantic_analyser.cpp
Comment (ajor) [inline src/ast/passes/semantic_analyser.cpp:395]: I'm guessing these bits of code are the reason for disallowing certain builtins in `check_available_in_subprog()`?

I think it'd be safer if we included the checks closer to the point where they're actually needed (here) - I can imagine in the future we'll forget about this implicit link, modify the list in `check_available_in_subprog` and end up with a SEGV from this dynamic_cast returning a null pointer.

i.e. something like:
```suggestion
    auto *probe = dynamic_cast<Probe *>(scope_);
    if (!probe)
    {
      LOG(ERROR, builtin.loc, err_) << builtin.ident << " cannot be used in functions";
      return;
    }
    for (auto &attach_point : *probe->attach_points)
```

**Review by ajor — CHANGES_REQUESTED** [review]
This is really looking pretty good now, thanks!

I just found some bugs that we should fix before merging.

Crash using disallowed builtins in functions:
```
$ sudo ./bpftrace -e 'fn foo(): int64 { return func; }' 
Segmentation fault
```

Crash writing to a map in a function:
```
$ sudo ./bpftrace -e 'fn bar(): void { @x = 1; }'
bpftrace: /home/ajor/src/bpftrace/src/ast/irbuilderbpf.cpp:444: void bpftrace::ast::IRBuilderBPF::CreateMapUpdateElem(llvm::Value*, bpftrace::ast::Map&, llvm::Value*, llvm::Value*, const bpftrace::location&): Assertion `ctx && ctx->getType() == getInt8PtrTy()' failed.
Aborted
```

Crash using map's value in a function:
```
$ sudo ./bpftrace -e 'fn bar(): uint64 { $res = @x + 1; return $res; } BEGIN { @x = 1; }'
bpftrace: /home/ajor/src/bpftrace/src/ast/irbuilderbpf.cpp:379: llvm::Value* bpftrace::ast::IRBuilderBPF::CreateMapLookupElem(llvm::Value*, bpftrace::ast::Map&, llvm::Value*, const bpftrace::location&): Assertion `ctx && ctx->getType() == getInt8PtrTy()' failed.
Aborted
```

Poor error message reading a map from a function:
```
$ sudo ./bpftrace -e 'fn bar(): int64 { return @x; } BEGIN { @x = 1; }'
stdin:1:20-29: ERROR: Function bar is of type int64, cannot return none
fn bar(): int64 { return @x; } BEGIN { @x = 1; }
                   ~~~~~~~~~
```

File: src/ast/passes/printer.cpp
Comment (lenticularis39) [reply src/ast/passes/printer.cpp:419]: Also I noticed the function outputs the old syntax of arguments:
```
# build/src/bpftrace -de 'fn test($x: int64): void {  } BEGIN {}'
[DEBUG] libbpf: loading kernel BTF '/sys/kernel/btf/vmlinux': 0
[DEBUG] libbpf: loading kernel BTF '/sys/kernel/btf/vmlinux': 0

AST after: parser
-------------------
Program
 test: void(int64 $x)
 BEGIN
```

File: src/ast/passes/codegen_llvm.cpp
Comment (lenticularis39) [reply src/ast/passes/codegen_llvm.cpp:2687]: Unlike probes which have a default return value, subprograms might leave unreachable blocks that are not finished with a return instruction, these have to be removed for the IR to be correct.

**Comment by lenticularis39** [comment]
Fixed maps and casts in subprogs. Also added more tests to cover this functionality.

Subprog argument implementation in codegen was changed to include ctx as the first argument, which is needed for perf-event-based builtin functions (like print) to work properly. This was tested at runtime with the in-progress function call PR.

With this all feedback should be resolved.

**Comment by danobi** [comment]
Are there any outstanding work items left? Or are we missing a final round of reviews?

Side note, this behavior is a bit unexpected:

```
$ sudo ./build/src/bpftrace -e 'fn foo($a: int32): int32 { return $a + 4; }'
stdin:1:30-43: ERROR: Function foo is of type int32, cannot return int64
fn foo($a: int32): int32 { return $a + 4; }

```

I suspect this is more related to how we represent all ints as int64 internally. Not a show stopper -- IMO fixing this relaxes some rules so this is ok to do later.

**Comment by danobi** [comment]
Also another bit of fix-later feedback:

```
sudo ./build/src/bpftrace -e 'fn foo(): void { return; }'
No probes to attach

$ sudo ./build/src/bpftrace -e 'fn foo() { return; }'
stdin:1:10-11: ERROR: syntax error, unexpected {, expecting :
fn foo() { return; }

```

Going with bpftrace's philosophy of brevity, perhaps good to allow `void` return types to be omitted. But I think this is debatable

**Comment by lenticularis39** [comment]
> Are there any outstanding work items left? Or are we missing a final round of reviews?
> 
No, just waiting for reviews.

> Side note, this behavior is a bit unexpected:
> 
> ```
> $ sudo ./build/src/bpftrace -e 'fn foo($a: int32): int32 { return $a + 4; }'
> stdin:1:30-43: ERROR: Function foo is of type int32, cannot return int64
> fn foo($a: int32): int32 { return $a + 4; }
>                              ~~~~~~~~~~~~~
> ```
> 
> I suspect this is more related to how we represent all ints as int64 internally. Not a show stopper -- IMO fixing this relaxes some rules so this is ok to do later.
> 
> Also the error underline seems to be off.

I know about this, it's because types were not explicitly given anywhere previously (besides casts). Not sure why the underlining is off.

File: src/ast/passes/codegen_llvm.cpp
Comment (ajor) [inline src/ast/passes/codegen_llvm.cpp:2431]: I'm assuming this is just a temporary thing until your next PR?

**Review by ajor — APPROVED** [review]
Sorry for taking so long to review this. Really appreciate the work you've put it!

The code's looking good to me - the comments are just a bunch of minor things that can be fixed later. Let's just get this one through without more delay!

The only issue I found was that we're allowing multiple functions to be defined with the same name. Again, this can be fixed later:
```
fn aaa(): void { print(1); }
fn aaa(): void { print(2); }
```

File: src/ast/passes/codegen_llvm.cpp
Comment (ajor) [reply src/ast/passes/codegen_llvm.cpp:2421]: Just been thinking about this some more and I realise that I don't actually understand. How can a subprogram generate an unreachable block without a return instruction?

File: src/ast/passes/semantic_analyser.cpp
Comment (lenticularis39) [reply src/ast/passes/semantic_analyser.cpp:2189]: Yes, the issue is this probably requires a new language feature, since in the context of a subprog, there is no way to tell if you are accessing userspace or kernel space data based on pointers alone.

File: src/ast/passes/codegen_llvm.cpp
Comment (lenticularis39) [reply src/ast/passes/codegen_llvm.cpp:2431]: This is because unlike with probes, there is not any default return value for subprogs. createRet() without argument is only used to end unreachable code that is then removed in subprogs.

File: src/ast/passes/codegen_llvm.cpp
Comment (lenticularis39) [reply src/ast/passes/codegen_llvm.cpp:2391]: Yeah I must have copied this from somewhere :D

File: src/ast/passes/codegen_llvm.cpp
Comment (lenticularis39) [reply src/ast/passes/codegen_llvm.cpp:881]: Currently subprogs do not track if they are called from an iterator probe so printf from subprog from iterator probe is treated like outside iterator probe. This should be probably fixed eventually.

**Comment by lenticularis39** [comment]
The PR should be now rebased and ready for merge.

**Comment by ajor** [comment]
Thanks @lenticularis39!

File: tests/semantic_analyser.cpp
Comment (ajor) [reply tests/semantic_analyser.cpp:1]: Added in #3062

**Pull request merged**
